File: pa2/Lex.c
// Jason Hetland
// jhetland
// pa2 Lex.c

#include <stdlib.h>
#include <string.h>
#include <stdio.h>

#include "List.h"

#define MAX_LEN 160

// Similar to insertion sort
List insertsort(char** array, int length){
  // makes a new list object
  List list = newList();

  // appends first element
  if (length > 0) 
     append(list, 0);
  for (int j = 1; j < length; j++){
    char *temp = array[j];
    int index = j - 1;
    //move cursor to current position
    moveTo(list, index);
    //sort
    while (index > -1 && strcmp(temp, array[getElement(list)]) < 1){
      index--;
      movePrev(list);
    }
    
    // if cursor is null, then prepend otherwise insert after 
    if(getIndex(list) == -1){
      prepend(list, j);
    } else {
      insertAfter(list, j);
    }
  }

  return list;
}

// implementation of strdup
char *strdup (const char *s) {
  char *d = malloc(strlen(s) + 1);
  if (d == NULL) return NULL;
  strcpy(d,s);
  return d;
}

//reads the file into an array of strings
char** read(FILE* input, int lines){
  char** array = malloc(sizeof(char**) * lines);
  char line[MAX_LEN];
  for(int i = 0; i < lines; i++){
    fgets(line, MAX_LEN, input);
    array[i] = strdup(line);
  }
  return array;
}

//frees all the malloced heap space used by the program
void freeall(char** lines, int lineCount, List list){
  for (int i = 0; i < lineCount; i++){
    free(lines[i]);
  }
  free(lines);
  freeList(&list);
}

int main(int argc, char* argv[]){
  int count = 0;
  char line[MAX_LEN];

  if(argc != 3) {
    printf("Invalid number of inputs");
    exit(1);
  }
  
  FILE* input = fopen(argv[1], "r");
  FILE* output = fopen(argv[2], "w");

  if(input == NULL){ 
    printf("Unable to open file %s for reading\n", argv[1]);
    return 1;
  } else if (output == NULL){
    printf("Unable to open file %s for reading\n", argv[2]);
    return 1;
  }

  // reads each line of the input file, then count and prints
  while( fgets(line, MAX_LEN, input) != NULL) {
    count++;
  }

  // closes and reopens file to reset fgets
  fclose(input);
  input = fopen(argv[1], "r");
  char** lines = read(input, count);

  // carries out insertion sort
  List list = insertsort(lines, count);

  // prints out the list by searching for indices in array
  for(moveTo(list, 0); getIndex(list) >= 0; moveNext(list)){
    fprintf(output, "%s", lines[getElement(list)]);
  } 

  //closes files
  fclose(input);
  fclose(output);
  freeall(lines, count, list);
  return(0);
}

File: pa1/Lex.java
//Jason Hetland
//Progamming Assignment 1 
//Lex.java. This is 

import static java.lang.System.*;
import java.io.*;
import java.util.Scanner;

class Lex{

    static Scanner open(String filename){
	Scanner scan = null;
	try{
	   scan = new Scanner(new File(filename));
	} catch (IOException except){
	   System.err.println("Invalid file");
	   exit(1);
	}
	return scan;
   }

     //This counts the number of lines in the file
   static int countLines(Scanner file){
	int index = 0;
	for(; file.hasNextLine(); file.nextLine()) index++;
	return index;
   }

     //Stores the lines in the array
   static String[] read(Scanner file, int lines){
	String[] array = new String[lines];
	for(int i = 0; file.hasNextLine(); i++){
	    array[i] = file.nextLine();
        }
	return array;
   }

    //similar to insertion sort
   static List insertionSort(String[] file){
	List list = new List();
      	//this inserts the first element into the list
	if(file.length > 0)
	list.append(0);
	//sorts
	for(int j = 1; j < file.length; j++){
	   String temp = file[j];
	   int index = j - 1;
	   //moves the cursor to the current position
	   list.moveTo(index);
	   while(index > -1 && temp.compareTo(file[list.getElement()]) < 
		1){
	  	index--;
		list.movePrev();
	   }

	//if the cursor == null then prepend
	   if(list.getIndex() == -1){
	      list.prepend(j);
	   } else{
	      list.insertAfter(j);
	   }
	}
	return list;
    }

    public static void main (String[] args){
	//checks for valid arguments
	if(args.length != 2){
	  System.err.println("Invalid Input");
	  exit(1);
       }

	//stores files in scanner objects
	Scanner input = open(args[0]);
	PrintWriter writer = null;
	try{
	   writer = new PrintWriter(new FileWriter(args[1]));
      } catch (FileNotFoundException excep){
	 System.err.println("Could not create the file");
	 exit(1);
      } catch (IOException excep){
	System.err.println("Could not create the file");
         exit(1);
      }

	// prints out the number of lines
	int lineCount = countLines(input);
  	// resets line Scanner and pointer
	input = open(args[0]);
	String[] array = read (input, lineCount);
	//stores indices in list
	List list = insertionSort(array);
	//writes the lines to the file
	for(list.moveTo(0); list.getIndex() >= 0; list.moveNext()){
	   writer.println(array[list.getElement()]);
	}
	
	writer.close();
   }
}
